### PR TITLE
refactor: move protocol out of rpc generation

### PR DESCRIPTION
**Pull request by imurchie**
Separate the protocol messages from the RPC generation, which is iOS specific. I've been working on integrating Chrome and this is a first step. Also makes the protocol easier to handle.

### CODE REVIEW

File: lib/protocol/index.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
+const OBJECT_GROUP = 'console';
+
+const COMMANDS = {
```
Comment (mykola-mokhnach) [inline lib/protocol/index.js:3]: I usually also add a link to the official doc for such declarations

File: lib/protocol/index.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  /* APPLICATIONCACHE DOMAIN */
```
Comment (mykola-mokhnach) [inline lib/protocol/index.js:4]: we could use the `region` notation to make sections easily collapsable in IDEs: https://code.visualstudio.com/updates/v1_17#_folding-regions

File: lib/protocol/index.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+  /* WORKER DOMAIN */
+  'Worker.enable': [],
+};
+
+function getCommand (id, method, params = {}) {
+  return {
+    id,
+    method,
+    params: Object.assign({
```
Comment (mykola-mokhnach) [inline lib/protocol/index.js:89]: using `Object.assign` will override default properties even if these contain null/undefined in `params`. Should we use lodash's `_.defaults` instead?

File: lib/protocol/index.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+function getProtocolCommand (id, method, opts = {}) {
+  const paramNames = COMMANDS[method];
+  if (!paramNames) {
+    throw new Error(`Unknown command: '${method}'`);
+  }
+
+  const params = paramNames.reduce(function (params, param) {
+    params[param] = opts[param];
```
Comment (mykola-mokhnach) [inline lib/protocol/index.js:107]: nit: it's a bit confusing to read if the result name is the same as the accumulator name. I'd rather rename the first arg of the reducer function to `acc` or `result`

File: lib/rpc/remote-messages.js [hunk 5]
```diff
@@ -1,8 +1,29 @@
 import _ from 'lodash';
+import getProtocolCommand from '../protocol';
 
 
 const OBJECT_GROUP = 'console';
 
+const MINIMAL_COMMAND = 'getMinimalCommand';
+const FUll_COMMAND = 'getFullCommand';
```
Comment (mykola-mokhnach) [inline lib/rpc/remote-messages.js:8]: perhaps `FULL_COMMAND`?

File: lib/rpc/remote-messages.js [hunk 6]
```diff
@@ -41,8 +60,7 @@ class RemoteMessages {
     };
   }
 
-  setSenderKey (opts = {}) {
-    const {connId, senderId, appIdKey, pageIdKey} = opts;
+  setSenderKey (connId, senderId, appIdKey, pageIdKey) {
```
Comment (mykola-mokhnach) [inline lib/rpc/remote-messages.js:63]: I liked the previous implementation more. For a client it's quite easier to mix arguments between each other if the count of these is greater than two

File: lib/rpc/rpc-client.js [hunk 7]
```diff
@@ -460,7 +460,9 @@ export default class RpcClient {
     await this.send('Memory.enable', sendOpts, false);
     await this.send('ApplicationCache.enable', sendOpts, false);
     await this.send('ApplicationCache.getFramesWithManifests', sendOpts, false);
-    await this.send('Timeline.setInstruments', sendOpts, false);
+    await this.send('Timeline.setInstruments', Object.assign({
```
Comment (mykola-mokhnach) [inline lib/rpc/rpc-client.js:463]: same comment as above

File: lib/rpc/remote-messages.js (hunk 6 above)
Comment (imurchie) [reply lib/rpc/remote-messages.js:63]: This is only called in one place, below. The overhead of creating a new object, etc, is not needed.

File: lib/rpc/rpc-client.js (hunk 7 above)
Comment (imurchie) [reply lib/rpc/rpc-client.js:463]: This involves nothing that is inputed from anywhere, so the caution is unnecessary.

**Pull request merged**